daml/ContingentClaims/Lifecycle.daml

```diff
@@ -42,56 +41,45 @@ data Result a = Result with
 -- 3. A set of `Claim`s.
 -- 4. A function which, given a date, lifecycles the third argument.
 lifecycle : (Eq a, CanAbort m)
-  => (Text -> Date -> m Decimal) 
-  -> (C a -> C a -> C a) 
-  -> C a 
-  -> Date -> m (Result a) 
+  => (Text -> Date -> m Decimal)
+  -> (C a -> C a -> C a)
+  -> C a
+  -> Date -> m (Result a)
 lifecycle spot choose
   = runKleisli
   . fmap (uncurry Result)
-  . flip runReaderT 1.0
+  . flip evalStateT 1.0
   . runWriterT
-  . para (lifecycle' spot choose)
-
-type Seed m a = WriterT [(Decimal, a)] (ReaderT Decimal (Kleisli m Date)) (C a)
-
--- TODO: these three should probably go in their respective modules
+  . apoCataM (pruneZeros') (lifecycle' spot choose)
 
-instance (Monoid w, CanAbort m) => CanAbort (WriterT w m) where
-  abort = lift . abort
-
-instance CanAbort m => CanAbort (ReaderT s m) where
-  abort = lift . abort
-
-instance CanAbort m => CanAbort (Kleisli m a) where -- because of the order of parameters it's not possible to write a MonadTrans instance for Kleisli.
-  abort = Kleisli . const . abort
+type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
 
 lifecycle' : (Eq a, Monad m, CanAbort m)
   => (Text -> Date -> m Decimal)
   -> (C a -> C a -> C a)
-  -> F a (C a, Seed m a) -> Seed m a
+  -> C a -> Seed m a
 
-lifecycle' _ _ ZeroF = pure Zero
-lifecycle' _ _ (OneF id) = do
-  qty <- ask
+lifecycle' _ _ Zero = pure ZeroF
+lifecycle' _ _ (One id) = do
+  qty <- get
   tell [(qty, id)]
-  pure $ One id
-lifecycle' _ _ (GiveF (_, m)) = local negate m
-lifecycle' spot _ (WhenF obs (c, m)) = do
-  wasLifecycled <- lift . lift $ eval spot obs
-  if wasLifecycled then m $> Zero else pure $ When obs c
-lifecycle' spot _ (ScaleF obs (_, m)) = do
+  pure ZeroF -- replace lifecycled node with zero, to avoid double-spend.
+lifecycle' _ _ (Give c) = do
+  modify negate 
+  pure $ GiveF (Right c)
+lifecycle' spot _ (When obs c) = do
+  predicate <- lift . lift $ eval spot obs
+  pure $ if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lifecycle' spot _ (Scale obs c) = do
   k <- lift . lift $ eval spot obs
-  local (* k) m
-lifecycle' _ choose (OrF (c, m) (c', m')) = do
+  modify (* k) 
+  pure $ ScaleF obs (Right c)
+lifecycle' spot choose (Or c c') = do
   let chosen = choose c c'
-  if | chosen == c -> m
-     | chosen == c' -> m'
+  if | chosen == c -> lifecycle' spot choose c --TODO: should we abort with nested choices?
```

Comment:
The comment here refers to the case where you have nested `Or`s. I haven't come across such a case yet. Also, because we're requiring the user to pass the entire subtree that would be chosen, it would fail (because the second invocation of `Or` would have to be smaller than the first one, so the argument will never be the same).
This is something we should try and figure out though. Sooner or later, somebody will come up with an instrument that has multiple choices.

Reply:
Why not change the signature of the `choose` function to `C a -> C a -> Bool`, to indicate whether the holder takes the first or second choice?

Reply:
A most excellent question! This is to ensure that the investor, off-ledger, has made a choice corresponding to the inputs of the function. If they were to return a boolean, there's no way to 'prove' that they chose the left or right branch, in case of a dispute.

Reply:
Why is there no way to prove it? The signatories of the contract see all of the inputs and outputs of lifecycling, which afaiu happens on-ledger. So `choose` is embedded in the Choice that calls lifecycle.

Reply:
So consider a real set-up where a user has a UI that they make their 'election' from. Now say their UI is buggy. The only thing we'll see as an input is `True` or `False` when the choice is exercised. If they say "but I chose to exercise my convertible bond!", and we see the opposite, it's hard to argue about who's right.
On the other hand, if they passed a claim, that structurally says `One Equity 'And' Give (One Bond)` then there's not much to dispute. 
Also, the likelihood of their buggy UI generating the correct tree 'randomly' is generally much lower. v.s. 'randomly' picking the wrong True/False is 1:2.

Reply:
A buggy UI is not an argument for how something should behave on the ledger, as UIs can be changed. The question of take the bond or take the cash is the same as do you want to take the first thing (the bond)? Framing the question as a `Bool` is better as then you can avoid having that `abort` to begin with!

Notice that in current implementation you're exposing an attach vector. Suppose within a Claim the choices are between two obligations `Give (Scale 100 (One USD))` or `Give (zcb ...)`,  I respond back with something else `Zero`, that is not one of the choices, lifecycling `Choice` aborts and we can never make progress on that contract/claim.